src/guest_memory.rs

```diff
@@ -681,10 +681,7 @@ pub trait GuestMemory {
     where
         F: ReadVolatile,
     {
-        self.try_access(count, addr, |offset, len, caddr, region| -> Result<usize> {
-            // Check if something bad happened before doing unsafe things.
-            assert!(offset <= count);
-
```

Comment:
How about completely remove the `offset` parameter, and change `try_access` accordingly.

In my view the only place might need the assertion is before the first execution of `f(total, len as usize, start, region)` in `try_access`, but since `total` defaults to 0 and `len` is an `usize`, the assertion there is pointless. The rest cases are well guarded in `Ok(len)` match arm inside `try_access`.

Reply:
The `offset` parameter is used in the `read` and `write` implementations in `Bytes<GuestAddress> for T: GuestMemory` :/

Well, technically, we cannot rely on the actual behavior of `try_access`, since it's a trait function, and some arbitrary implementation of `GuestMemory` could overwrite it with some absolute nonsense. The main point for me here is that even in that case, it will not cause unsoundness, because there's no unsafe code here, and all bounds are properly checked further down the stack.

Reply:
> The `offset` parameter is used in the `read` and `write` implementations in `Bytes<GuestAddress> for T: GuestMemory` :/

Oh I didn't read that far, let's just remove the `assertion` here then 😉